crates/sui-source-validation/src/lib.rs

```diff
@@ -0,0 +1,238 @@
+    /// A local dependency was not found
+    LocalDependencyNotFound {
+        package: Symbol,
+        module: Option<Symbol>,
+    },
+    /// An on-chain version of a local dependency was not found
+    ///
+    /// params:  package, module
+    OnChainDependencyNotFound(Symbol, Option<Symbol>),
+    /// Local dependencies have a different number of modules than on-chain
+    ///
+    /// params:  expected count, on-chain count, missing
+    ModuleCountMismatch(usize, usize, Vec<String>),
+    /// A local dependency module did not match its on-chain version
+    ///
+    /// params:  package, module, address
+    ModuleBytecodeMismatch(String, String, AccountAddress),
+}
+
+impl Display for DependencyVerificationError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        Debug::fmt(&self, f)
+    }
+}
+
+#[derive(Debug)]
+pub struct BytecodeSourceVerifier<'a> {
+    pub verbose: bool,
+    rpc_client: &'a ReadApi,
+}
+
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub struct Dependency {
+    pub symbol: String,
+    pub module_bytes: BTreeMap<String, Vec<u8>>,
+}
+
+/// map the package's direct dependency symbols to their addresses and nested dependencies
+type ModuleBytesMap = HashMap<Symbol, (AccountAddress, HashMap<Symbol, Vec<u8>>)>;
+
+impl<'a> BytecodeSourceVerifier<'a> {
+    pub fn new(rpc_client: &'a ReadApi, verbose: bool) -> Self {
+        BytecodeSourceVerifier {
+            verbose,
+            rpc_client,
+        }
+    }
+
+    /// Verify that all local Move package dependencies' bytecode matches
+    /// the bytecode at the address specified on the Sui network we are publishing to.
+    pub async fn verify_deployed_dependencies(
+        &mut self,
+        compiled_package: CompiledPackage,
+    ) -> Result<(), DependencyVerificationError> {
+        let compiled_dep_map = get_module_bytes_map(&compiled_package);
+
+        let mut on_chain_module_count = 0usize;
+        let mut verified_dependencies: HashMap<AccountAddress, Dependency> = HashMap::new();
+
+        for (pkg_symbol, (addr, local_pkg_bytes)) in compiled_dep_map {
+            if pkg_symbol == compiled_package.compiled_package_info.package_name {
+                continue;
+            };
+
+            // fetch the Sui object at the address specified for the package in the local resolution table
+            let on_chain_package = self.pkg_for_address(&addr).await?;
+
+            for (module_symbol, local_bytes) in local_pkg_bytes {
+                // package addresses may show up many times, but we only need to verify them once
+                // zero address is the package we're checking dependencies for
+                if verified_dependencies.contains_key(&addr) || addr == AccountAddress::ZERO {
+                    continue;
+                }
+
+                let mod_str = module_symbol.to_string();
+                let on_chain_bytes = match on_chain_package.module_map.get(&mod_str) {
```

Comment:
You can use `let-else` now! Hurray! `let Some(on_chain_bytes) = on_chain_package.module_map.get(&mod_str) else { return Err(DependencyVerificationError::OnChainDependencyNotFound(pkg_symbol,Some(module_symbol)))`

Not sure its that much better. I'm just really excited by `let-else` :D